# stochastic-regulation-code
Code for the manuscript "Inferring gene regulation from stochastic transcriptional variation across single cells at steady-state"

# Abstract
Regulatory relationships between transcription factors (TFs) and their target genes lie at the heart of cellular identity and function; however, uncovering these relationships is often labor-intensive and requires perturbations. Here, we propose a principled framework to systematically infer gene regulation for all TFs simultaneously in cells at steady-state by leveraging the intrinsic variation in the transcriptional abundance across single cells. Through modeling and simulations, we characterize how transcriptional bursts of a TF gene are propagated to its target genes, including the expected ranges of time delay and magnitude of maximum covariation. We distinguish these temporal trends from the time-invariant covariation arising from cell states, and we delineate the experimental and technical requirements for leveraging these small but meaningful co-fluctuations in the presence of measurement noise. While current technology does not yet allow adequate power for definitively detecting regulatory relationships for all TFs simultaneously in cells at steady-state, we investigate a small-scale dataset to inform future experimental design. This study supports the potential value of mapping regulatory connections through stochastic variation, and it motivates further technological development to achieve its full potential.

# This repo contains:
- the transcriptional bursting + regulation simulator code (stochastic_bursting_simulatory.py)
- code to run the simulator on command line (run_simulator.py): please edit the gene-specific parameter values and number of simulation runs based on what you would like to explore
- code to run the simulated data analyses included in the manuscript (simulated_data_analyses_Apr2022.ipynb): please note that the simulated data files used here can be re-generated by running the simulator with the appropriate parameter settings
- code to process the K562 scRNA-seq data from the time-resolved experiment of cells at steady-state: the data have been deposited to GEO (GSE202292).
  - Scripts to preprocess raw data: 
  - R based code to generate panels 4B, 4C, S4A, S4G and S4H: stochastic_regulation_figure_4_S4_code_R.ipynb
  - R based code generate panel 4C:
  - Python based code to generate panels S4C, S4D, S4E, S4F: 
